### PR TITLE
Add asset accessor to get asset info from asset server

### PR DESCRIPTION
**Pull request by stark-dev**
Accessor class to get data by interacting directly with asset server (PoC)

Implements only one method to get the asset database ID from the internal name

Signed-off-by: Mauro Guerrera <mauroguerrera@eaton.com>

### CODE REVIEW

**Review by perrettecl — APPROVED** [review]

File: accessor/public_includes/fty_asset_accessor.h [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+namespace fty
+{
+
+    class AssetAccessor
+    {
+    public:
+        AssetAccessor();
+        ~AssetAccessor();
+
+        uint32_t assetInameToID(const std::string &iname);
```
Comment (eric3873) [inline accessor/public_includes/fty_asset_accessor.h:18]: could be static ?

**Review by eric3873 — APPROVED** [review]

File: accessor/CMakeLists.txt [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+##############################################################################################################
+find_package(fty-cmake PATHS ${CMAKE_BINARY_DIR}/fty-cmake)
+##############################################################################################################
+
+
+##############################################################################################################
+
+etn_target(shared ${PROJECT_NAME}
+    SOURCES
+        src/*.cc
```
Comment (etnKlendathu) [inline accessor/CMakeLists.txt:17]: please, better use file names then mask.

File: accessor/public_includes/fty_asset_accessor.h [hunk 3]
```diff
@@ -0,0 +1,21 @@
+#pragma once
+
+#include <fty_asset_dto.h>
+#include <fty_common_client.h>
+#include <fty_common_mlm_sync_client.h>
+#include <memory>
+#include <string>
```
Comment (etnKlendathu) [inline accessor/public_includes/fty_asset_accessor.h:7]: are those includes needed here? :)

File: accessor/public_includes/fty_asset_accessor.h [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+
+namespace fty
+{
+
+    class AssetAccessor
+    {
+    public:
+        AssetAccessor();
+        ~AssetAccessor();
```
Comment (etnKlendathu) [inline accessor/public_includes/fty_asset_accessor.h:16]: Are default ctor, dtor is needed?

File: accessor/src/fty_asset_accessor.cc [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+
+        msg.metaData().emplace(messagebus::Message::CORRELATION_ID, messagebus::generateUuid());
+        msg.metaData().emplace(messagebus::Message::SUBJECT, "GET_ID");
+        msg.metaData().emplace(messagebus::Message::FROM, ACCESSOR_NAME);
+        msg.metaData().emplace(messagebus::Message::TO, ASSET_AGENT);
+        msg.metaData().emplace(messagebus::Message::REPLY_TO, ACCESSOR_NAME);
+
+        msg.userData().push_back(iname);
+
+        messagebus::Message ret = interface->request("FTY.Q.ASSET.QUERY", msg, 10);
```
Comment (etnKlendathu) [inline accessor/src/fty_asset_accessor.cc:40]: try/catch?

File: accessor/src/fty_asset_accessor.cc [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+
+        log_debug("Received: %s", ret.userData().front().c_str());
+
+        if(ret.metaData().at(messagebus::Message::STATUS) != messagebus::STATUS_OK) {
+            return 0;
+        }
+        return static_cast<uint32_t>(std::stoul(ret.userData().front()));
```
Comment (etnKlendathu) [inline accessor/src/fty_asset_accessor.cc:47]: fty::convert<uint32_t>(ret.userData().front())

File: accessor/src/fty_asset_accessor.cc [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+        msg.metaData().emplace(messagebus::Message::REPLY_TO, ACCESSOR_NAME);
+
+        msg.userData().push_back(iname);
+
+        messagebus::Message ret = interface->request("FTY.Q.ASSET.QUERY", msg, 10);
+
+        log_debug("Received: %s", ret.userData().front().c_str());
+
+        if(ret.metaData().at(messagebus::Message::STATUS) != messagebus::STATUS_OK) {
+            return 0;
```
Comment (etnKlendathu) [inline accessor/src/fty_asset_accessor.cc:45]: Look into fty::expected to return error or value...

File: accessor/public_includes/fty_asset_accessor.h (hunk 1 above)
Comment (stark-dev) [reply accessor/public_includes/fty_asset_accessor.h:18]: Sure, changing it ;)

File: accessor/public_includes/fty_asset_accessor.h (hunk 3 above)
Comment (stark-dev) [reply accessor/public_includes/fty_asset_accessor.h:7]: Just string probably, copying and paste sorry

File: accessor/public_includes/fty_asset_accessor.h (hunk 4 above)
Comment (stark-dev) [reply accessor/public_includes/fty_asset_accessor.h:16]: nope, removing it

File: accessor/public_includes/fty_asset_accessor.h [hunk 8]
```diff
@@ -0,0 +1,38 @@
...
+#include <fty/expected.h>
+#include <string>
+
+namespace fty
+{
+
+    class AssetAccessor
+    {
+    public:
+        AssetAccessor() = default;
```
Comment (etnKlendathu) [inline accessor/public_includes/fty_asset_accessor.h:33]: you dont need ctor at all... you have a class with 1 static method. IMHO this looks wierd. :)

File: accessor/src/fty_asset_accessor.cc [hunk 9]
```diff
@@ -0,0 +1,75 @@
...
+
+        try
+        {
+            ret = interface->request(ASSET_AGENT_QUEUE, msg, RECV_TIMEOUT);
+            log_debug("assetInameToID received: %s", ret.userData().front().c_str());
+        }
+        catch (messagebus::MessageBusException &e)
+        {
+            log_error("MessageBus request failed: %s", e.what());
+            return fty::unexpected("MessageBus request failed");
```
Comment (etnKlendathu) [inline accessor/src/fty_asset_accessor.cc:64]: maybe is good idea to provide exception what() to return error.

File: accessor/src/fty_asset_accessor.cc [hunk 10]
```diff
@@ -0,0 +1,75 @@
...
+        msg.metaData().emplace(messagebus::Message::REPLY_TO, ACCESSOR_NAME);
+
+        msg.userData().push_back(iname);
+
+        messagebus::Message ret;
+
+        try
+        {
+            ret = interface->request(ASSET_AGENT_QUEUE, msg, RECV_TIMEOUT);
+            log_debug("assetInameToID received: %s", ret.userData().front().c_str());
```
Comment (etnKlendathu) [inline accessor/src/fty_asset_accessor.cc:59]: I don't know, just for thinking about logging in aux libraries... you use a log here, but not manage logging at all...

File: accessor/test/test.cpp [hunk 11]
```diff
@@ -0,0 +1,49 @@
...
+        AssetAccessor accessor;
+    }());
+}
+
+TEST_CASE("Request ID - Success")
+{
+    auto id = AssetAccessor::assetInameToID("rackcontroller-0");
+    
+    CHECK(id);
+    CHECK(*id == 1);
```
Comment (etnKlendathu) [inline accessor/test/test.cpp:38]: CHECK(id); -> check a value and continue to execute this test
CHECK(*id == 1); if !id -> assert

So, better to use REQUIRE in first line to break this test execution

File: accessor/test/test.cpp [hunk 12]
```diff
@@ -0,0 +1,49 @@
...
+}
+
+TEST_CASE("Request ID - Failure")
+{
+    AssetAccessor accessor;
+    auto id = AssetAccessor::assetInameToID("datacenter-0");
+
+    CHECK(!id);
+    CHECK(id.error() == "Request of ID from iname failed");
```
Comment (etnKlendathu) [inline accessor/test/test.cpp:47]: the same

File: accessor/src/fty_asset_accessor.cc [hunk 13]
```diff
@@ -0,0 +1,72 @@
...
+
+        messagebus::Message ret;
+
+        try
+        {
+            ret = interface->request(ASSET_AGENT_QUEUE, msg, RECV_TIMEOUT);
+        }
+        catch (messagebus::MessageBusException &e)
+        {
+            return fty::unexpected(std::string("MessageBus request failed: ") + e.what());
```
Comment (etnKlendathu) [inline accessor/src/fty_asset_accessor.cc:61]: Hehe... 
return fty::unexpected("MessageBus request failed: {}", e.what());

**Review by etnKlendathu — APPROVED** [review]
Approved with a smaaaall comment

File: accessor/src/fty_asset_accessor.cc (hunk 13 above)
Comment (stark-dev) [reply accessor/src/fty_asset_accessor.cc]: fixed :)

**Review by perrettecl — APPROVED** [review]

**Review by perrettecl — APPROVED** [review]

**Pull request merged**